src/api.ts

```diff
@@ -0,0 +1,50 @@
+import StoryblokClient from 'storyblok-js-client'
+import { session } from './session'
+
+export interface ApiClientState {
+  region: string
+  accessToken: string
+  client: StoryblokClient | null
+}
+
+const state: ApiClientState = {
+  region: 'eu',
+  accessToken: '',
+  client: null,
+}
+
+export function apiClient() {
```

Comment:
I think OOP in JS is getting ditched in favor of FP, but I also agree with @Edo-San that this style is more suitable for OOP, I just checked this as "personal preferences". Should we enforce a code style even for such cases?

I saw other lines where I would have preferred a different style (like ternary or simplifications in general) but I thought that in the end this is a choice, and every dev has its style: we enforce the syntax for obvious reasons and a macro approach for consistency, but regarding individual choices in code writing I don't know.

Reply:
I'm going to tackle the OPP vs Procedural topic here @Edo-San @edodusi, I'm happy this topic is raised now because it's important to tackle.

@Edo-San @edodusi  do you have any javascript project where you decided on OOP instead of functional programming so I can understand the benefits of using in the context of this kind of tools? That would be helpful.

This is a topic I'm particularly very opinionated on but I'm open to understanding every point of view. Based on my experience, I can't count the times that teams I was involved with needed to refactor legacy code based on classes:

Javascript is inherently prototype-based, it's based on functions and objects. Classes in Javascript are purely a sugar syntaxis because before ES5, most people [didn’t take JS seriously as a language](https://medium.com/@_benaston/lesson-1a-the-history-of-javascript-8c1ce3bffb17). The inclusion of class syntax in JS was an intentional decision to attract these more experienced programmers and lower the barrier of entry to make JS more popular for devs coming from Java and C#, not because it has a significant benefit.

That being said, there are use cases where I found classes and OOP more suitable, for example: Three.js works with classes under the hood, because the nature of 3D objects and materials can benefit from this pattern, a ShaderMaterial inherits and extends the `Material` class which inherits `Object3D` etc. It is also useful for game development, like using Phaser because games are essentially OOP.

But in this particular case, we are creating a CLI tool that is, by nature, procedural, executes a series of commands, and any need of a class like session class can be easily translated to singleton functions, procedural programming, focusing on functions and actions, fits this model naturally. No inheritance or performance need justifies OOP. Additionally, functional programming brings other benefits that are well-suited to a CLI tool, such as immutability, pure functions, and higher-order functions and the list go on.

If Typescript is the argument in favor, we can achieve the same type-safety using interfaces and function-based patterns. We can use interfaces or types to define the shape of objects and functions.

Another reason for me to continue with the functional approach is to use the struggles I found in increasing the testing coverage of the `storyblok-js-client` as an example of why we shouldn't be using classes:

If you check the code, you can find that all the utils and helpers are using a class that is used only twice. https://github.com/storyblok/storyblok-js-client/blob/ea688457e6b0706b5b88553e0a8e6fa5e95ee14a/src/sbHelpers.ts#L12. This made unit testing and mocking a nightmare where just by simply having anonymous functions without side-effects that are easily testable would have been enough. And I can't expose the private properties to be able to test because that would be a breaking change... Same for the sbFetch class, instead of using something like ofetch.

My goal is that our tooling is flexible, easy to test, maintainable, and scalable, and functional programming is key for this, so unless there is a solid reason to use OOP on our tools, I feel more comfortable by delivering tools with a functional approach, which is important as well, because is going to be 2/3 of my work here.

We can also take the discussion to the call as well.

Reply:
@alvarosabu maybe I was not clear: I'm in favor of FP, not OOP. The matter here and I think @Edo-San was referring to this one is that you are using nested functions, which is a code style choice, and the code would maybe be more readable using OOP in this specific case.

And my take on this is: it's a personal choice. You like nested functions, I usually don't, but it's not something I would enforce, as the code is perfectly understandable.

Reply:
Hey @edodusi we discussed the topic over a call and we decided on the following:

OOP vs Functional Programming: we decided to go for "the best tool to do the work" meaning we only use classes in the ecosystem when they are beneficial (for example, inheritance or polymorphism), and in the specific case of the CLI, functional programming has major pros, especially regarding mocking and testing.

> And my take on this is: it's a personal choice. If you like nested functions, I usually don't, but it's not something I would enforce, as the code is perfectly understandable.

I differ from you about being a style choice, javascript is based on functions and objects, if you use classes you also use nested functions that have a sugar syntax on top of it, as far as I discussed this with Edo, he mentioned that some functions (like session) **felt** like they could be classes, we just discussed the pros and cons and in this particular case, there is nothing that justifies using classes apart of having custom Error classes extending the original one just because of the polymorphism.

Reply:
Im gonna resolve this since we got an agreement on Slack